File: API/src/main/java/com/aklaa/api/services/contract/GroceryListService.java
package com.aklaa.api.services.contract;

import com.aklaa.api.dtos.request.CartDishRequestDTO;
import com.aklaa.api.dtos.request.GroceryListIngredientListRequestDTO;
import com.aklaa.api.dtos.response.GroceryListIngredientListResponseDTO;
import com.aklaa.api.dtos.response.GroceryListResponseDTO;
import com.aklaa.api.model.User;
import jakarta.servlet.http.HttpSession;
import org.springframework.data.domain.Pageable;

import java.util.List;

/**
 * Service interface for grocery list and shopping cart management operations.
 * <p>
 * This service provides functionality for managing shopping carts in sessions,
 * retrieving grocery lists, and managing ingredients within grocery lists.
 * </p>
 */
public interface GroceryListService {
    
    /**
     * Retrieves the shopping cart from the HTTP session.
     * <p>
     * This method fetches the cart stored in the session under the "cart" attribute.
     * If the cart doesn't exist or is invalid, a new empty cart is created and stored in the session.
     * </p>
     *
     * @param session the HTTP session containing the shopping cart
     * @return a list of {@link CartDishRequestDTO} representing the dishes in the cart
     */
    List<CartDishRequestDTO> getCart(HttpSession session);
    
    /**
     * Retrieves all grocery lists for a specific user with pagination.
     * <p>
     * This method fetches grocery lists belonging to the user, sorted and paginated
     * according to the provided pageable parameters.
     * </p>
     *
     * @param user the user whose grocery lists should be retrieved
     * @param pageable the pagination information including page number, size, and sorting
     * @return a list of {@link GroceryListResponseDTO} containing the user's grocery lists
     */
    List<GroceryListResponseDTO> getGroceryLists(User user, Pageable pageable);
    
    /**
     * Retrieves the ingredients of a specific grocery list with pagination.
     * <p>
     * This method fetches all ingredients in a grocery list, sorted by category and name,
     * and returns them in a paginated format. If the grocery list is not found or doesn't
     * belong to the user, an empty response is returned.
     * </p>
     *
     * @param id the ID of the grocery list
     * @param user the user who owns the grocery list
     * @param pageable the pagination information including page number, size, and sorting
     * @return a {@link GroceryListIngredientListResponseDTO} containing the paginated ingredients,
     *         total elements, and total pages
     */
    GroceryListIngredientListResponseDTO getIngredientOfGroceryList(Long id, User user, Pageable pageable);
    
    /**
     * Updates the ingredients in a specific grocery list.
     * <p>
     * This method updates ingredient quantities, removes ingredients not in the update list,
     * and adds new ingredients. Existing ingredients have their quantities updated,
     * while ingredients not in the request are removed from the list.
     * </p>
     *
     * @param id the ID of the grocery list to update
     * @param list the request DTO containing the updated ingredient quantities
     * @param user the user who owns the grocery list
     */
    void updateIngredientsOfGroceryList(Long id, GroceryListIngredientListRequestDTO list, User user);
}

File: Web/src/services/groceryListService.js
import api from '../api/axiosConfig';

/**
 * Grocery List Service
 * Handles all grocery list-related API calls
 */

const groceryListService = {
  /**
   * Get all grocery lists
   * @returns {Promise} All grocery lists
   */
  getGroceryLists: async () => {
    const response = await api.get('/grocerylists', {});
    return response;
  },

  /**
   * Get grocery list ingredients by ID
   * @param {number|string} id - Grocery list ID
   * @param {Object} params - Additional parameters
   * @returns {Promise} Grocery list ingredients
   */
  getGroceryListIngredients: async (id, params = {}) => {
    const response = await api.get(`/grocerylists/${id}/ingredients`, { params });
    return response;
  },

  /**
   * Update grocery list
   * @param {number|string} id - Grocery list ID
   * @param {Object} data - Updated grocery list data
   * @returns {Promise} Updated grocery list
   */
  updateGroceryList: async (id, data) => {
    return api.put(`/grocerylists/${id}`, data);
  },

  /**
   * Delete grocery list
   * @param {number|string} id - Grocery list ID
   * @returns {Promise} Delete response
   */
  deleteGroceryList: async (id) => {
    return api.delete(`/grocerylists/${id}`);
  },

  /**
   * Save cart to grocery list
   * @param {Object} params - Save parameters
   * @returns {Promise} Save response
   */
  saveCartToGroceryList: async (params = {}) => {
    return api.post('/grocerylists/save', null, { params });
  },
};

export default groceryListService;
